hardening guide: incorporate topics from SAP HANA hardening guide

mgerstner opened the pull request:
### Description

This adds two new sections to the SLE hardening guide that originate from the SAP HANA hardening guide. The topics aren't specific to SAP. A customer complained about lacking aspects in these two sections. The concerns have been addressed and the content updated and adjusted to the more general audience of the SLE hardening guide.


### Are there any relevant issues/feature requests?

* bsc#1174501


### Is this feature exclusive to SLE 15 SP3 (or Leap 15.3)?

No

- [ ] This PR only applies to SLE 15 SP3 or higher
- [X] This PR applies to older releases as well.


### Are backports required?

- [ ] To maintenance/SLE15SP2
- [ ] To maintenance/SLE15SP1
- [ ] To maintenance/SLE15SP0
- [x] To maintenance/SLE12SP5
- [X] To maintenance/SLE12SP4
- [ ] To maintenance/SLE12SP3
- [ ] To maintenance/SLE12SP2

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+     We recommend to use the <literal>screen</literal> tool in order to
+     detach sessions before logging out. Screen sessions are not terminated
+     and can be re-attached whenever it is required. Active session can be
+     locked without logging out (read about <literal>C-a x</literal> /
```
```suggestion
     locked without logging out (read about <keycombo action="seq">
      <keycombo action="press">
       <keycap function="control"/>
      <keycap>a</keycap>
      </keycombo>
      <keycap>x</keycap>
     </keycombo> /
```

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+   <para>
+    By default the <literal>root</literal> user is assigned a password
+    and can login on any local terminal, in a graphical session or via SSH.
+    For security sensitive systems <literal>root</literal> logins should be
```
```suggestion
    For security sensitive systems &rootuser; logins should be
```

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+    restricted to as few methods as possible. Shared usage of the root account
+    should be avoided, instead individual administrators should use tools like
+    <literal>su(1)</literal> or <literal>sudo(8)</literal> to elevate privileges.
+    This allows to associate <literal>root</literal> logins with particular users.
```
```suggestion
    This allows to associate &rootuser; logins with particular users.
```

tomschr commented:
Thank you @mgerstner. You're feedback is very welcome.  :+1: You will get feedback soon, I'm almost finished.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+     systems via input devices connected to a KVM switch or a remote
+     management card (ILO, DRAC, etc).
+     By default, Linux offers 6 different consoles, that can be switched to
+     via the key-combinations Alt+F1 - Alt+F6, when running in text mode, or
+     Ctrl+Alt+F1 - Ctrl+Alt+F6 when running in a graphical session. The
```
and there I thought Docbook couldn't become any more bloated ;-)

tomschr replied:
Haha, I know, I know. :smile: We need it for consistency reasons. The above markup allows us to render them appropriately (gray background, borders etc). Without it, that rendering won't be applied to pure text.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+     never be used for everyday system administration tasks.
+    </para>
+
+    <itemizedlist mark="bullet" spacing="normal">
```
These defaults are already used throughout the document. Should I remove them all then?

tomschr replied:
Yes please! :+1:

mgerstner commented:
I have rebased the two existing commits to adjust the style according to your comments. Regarding the `itemizedlist` defaults I've removed them from the whole document, to make it consistent, in a separate commit.

tomschr commented:
@mgerstner Thank you Matthias! Perfect! :+1:

If you don't object or need more time, I will do the second round and look into text issues.

mgerstner commented:
> If you don't object or need more time, I will do the second round and look into text issues.

Of course, just go ahead.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     </listitem>
+     <listitem>
+      <para>
+       This will include the <literal>pam_securetty</literal> PAM module during the
+       authentication process on local consoles to allow root only to login on
+       tty devices that are listed in the file <filename>/etc/securetty</filename>.
```
This is an explanation, not an action. As such, I think this belongs to the first step. I'd suggest to move that as a para into the first step (after the `<screen>` as an additional `<para>`).

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+       This will include the <literal>pam_securetty</literal> PAM module during the
+       authentication process on local consoles to allow root only to login on
+       tty devices that are listed in the file <filename>/etc/securetty</filename>.
+       Thus to actually limit the tty devices for root you need to remove all
+       entries from this file except one:
```
Try to start with a verb:

```suggestion
       Remove all entries from this file except one. This limits the access to TTY devices for root.
```

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     </listitem>
+     <listitem>
+      <para>
+       After these changes root logins from non-allowed terminals will be
+       rejected right away without even querying the root password.
```
I'm not sure about this. What should the user do here? It sounds like it's pure information? If yes, then it should be added after the action. Maybe add it to the last step after the `<screen>`?. 

I would create a different action here. Something like "Check bla..." to see if everything is okay. Does that make sense?

mgerstner replied:
Makes sense. I'll change it into a verification step.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+
+    <important>
+     <para>
+      One might be tempted to add the <literal>pam_securetty</literal> PAM
+      module to the <filename>/etc/pam.d/common-auth</filename> file to make it
+      effective for all PAM services. This would break <literal>su</literal>
+      and <literal>sudo</literal>, however, because these tools would then
```
Maybe simplify it?

```suggestion
      Avoid adding the  <literal>pam_securetty</literal> PAM  module to the
      <filename>/etc/pam.d/common-auth</filename> file. This change would
      break the <command>su</command> and <command>sudo</command>
      commands as the reject &rootuser; authentication.
```

Just an idea. :wink:

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     </para>
+    </important>
+
+    <important>
```
If possible, try to avoid three consecutive admonitions (2x important and note). This kind of "dilutes" the warning. Could you combine the two important into one?

mgerstner replied:
A lot of admonitions may seem strange but these are facts. Combining the two important blocks is technically possible, of course, but they are still two distinct items. So somebody might read the first few sentences and think "doesn't affect me" and skip over the rest of the block.

tomschr replied:
I was more speaking from a documentation and readability perspective. If you think this is too hard then leave it as it is for the time being. :-)

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     </para>
+    </important>
+
+    <note>
```
I think, this note is essential for the procedure above. Maybe move that *before* the procedure?

mgerstner replied:
fine with me, I'll move if up

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+    <title>Restricting Graphical Session Logins</title>
+
+    <para>
+     On security sensitive systems like servers it is best not to run a
+     graphical user environment at all. If you require a graphical login
+     anyways then it is even more so important that the &rootuser;
+     user cannot login in a graphical session. Graphical programs are often
+     not designed to be run as &rootuser; and are more likely to
+     contain security issues than console programs.
```
"On security sensitive systems..." same here. I'd took the liberty here to rephrase it a bit more.

```suggestion
    To improve security on your server, avoid to use
    graphical user environments at all.
    Graphical programs are often not designed to be run as
    &rootuser; and are more likely to contain security issues than
    console programs.
    If you require a graphical login, use a non-&rootuser; login.
    Configure your system to disallow &rootuser; login on graphical
    sessions.
```

I've structured it as follows: what is the goal ("to improve security"), then an explanation (the reason), and after that how to avoid it.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     not blocked by the firewall). To restrict this possibility the following
+     change of the OpenSSH configuration needs to be performed:
+    </para>
+    <itemizedlist>
```
Same here, also use a procedure/step structure.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+    </itemizedlist>
+    <note>
+     <para>
+      Using the PAM module <literal>pam_securetty</literal> is not suitable in
+      case of OpenSSH, because SSH logins not necessarily go through the PAM
+      stack during the auth phase (e.g. when using SSH public key authentication).
+      Also an attacker could be able to differentiate between a wrong password
+      and a successful login that was only rejected later on by policy.
```
This is not an action, but information. Is there anything the user has to do?

mgerstner replied:
No, that's why this is a `<note>` ;)

tomschr replied:
Ok, fine with me. :-)

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+  <sect1 xml:id="sec-sec-prot-inactivity-logout">
+   <title>Setting an Inactivity Timeout for Interactive Shell Sessions</title>
+   <para>
+    It can be a good idea to terminate an interactive shell session after a
+    certain time of inactivity. For example an administrator might have
+    forgotten to terminate a local login session on a sensitive system. This
+    could provide an attack vector if a local attacker finds such a session
+    unguarded. Another effect of idle shell sessions is waste of system
+    resources.
```
Maybe use a list to present the reasons why it is a good idea to terminate an interactive shell session. It could look like this:

```
It can be a good idea to terminate an interactive shell session after a certain time
of inactivity, for example:

* Prevents open, unguarded sessions on sensitive systems.
* Avoids waste of system resources.
```

Maybe it has to be phrased a bit better, but I hope you get the idea. :wink:

mgerstner replied:
Okay. But ... aren't all systems sensitive? :-P

tomschr replied:
Hehe, that's an almost philosophical question. ;-) It probably depends on how you define "sensitive". ;-)

tomschr replied:
Hmn... if *all* systems are sensitive, then what's the meaning of insensitive systems? :wink:  To me it raises more questions than that the term solves. Either define what you mean by "sensitive systems" or remove it.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+    resources.
+   </para>
+   <para>
+    By default there is no inactivity timeout for shells, so nothing will
```
```suggestion
    By default, there is no inactivity timeout for shells. Nothing will
```

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+    By default there is no inactivity timeout for shells, so nothing will
+    happen if a shell stays open unused for days or even years. It is possible
+    to configure most shells in a way that idle sessions terminate
+    automatically after a certain amount of time. In the following example we
+    will show how to set an inactivity timeout for a number of
```
To avoid "we" and to make it consistent with the other text:

```suggestion
    automatically after a certain amount of time. The following example
    shows how to set an inactivity timeout for a number of
```

tomschr reviewed (changes requested):
@mgerstner Thanks Matthias for all your work. Much appreciated! :+1: 

I've went through your text and added some comments. Hope I didn't misread your text. In case you have any questions, just let me know.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+   <para>
+    By default the &rootuser; user is assigned a password
+    and can login on any local terminal, in a graphical session or via SSH.
+    For security sensitive systems &rootuser; logins should be
+    restricted to as few methods as possible. Shared usage of the root account
```
> "For security sensitive systems"... that sounds like there are systems which don't need security. I think, this gives the wrong impression. We should tell our customers that security is _always_ important.

I see what you're getting at. But there is a difference between, say, a workstation in a restricted office, a server in a data center and a laptop used in the field. It could also raise the question why these hardenings need to be applied manually then and aren't already setup by default ;-). In this light I think it is legit to notice that these hardenings are not for everybody but for a specific target audience that worry especially about security or have especially sensitive setups. But I'm still okay to drop this part.

> Another issue is, I'm not sure what you mean by "methods". Accounts? As such, I'd suggest to rephrase it like:

Actually, no. The methods refer to the "local terminal", "graphical session" and "SSH", different ways to login as root on a system.

tomschr replied:
Ahh, I see what you mean. 🙂 Is that a common term? I heard it before. It could make sense to little bit elaborate on that. It doesn't need much, but maybe an additional sentence would be fine. 😉

Oh, then I misread your sentence, sorry. Well, in that case would it make sense to add an additional hint? Something like this:

    logins should be restricted to as few methods (like local terminal, SSH, graphical sessions, etc.) 
    as possible.

mgerstner replied:
I think that would be strangely redundant. I try a bit of a different approach to make it more clear.

You mean if "login methods" is a common term? Well Google knows a couple of million of occurrences ;-). Apart from that I don't think the term "method" is uncommon in technical contexts.

tomschr replied:
Hmn, yes. Ok, I'll leave that to Liam, maybe he has some idea.

lproven replied:
I think that "login method" is a valid phrase, but ambiguous. I would just suggest slightly clarifying the list of examples:

> logins should be restricted to as few methods (such as on the local console, via SSH, graphical sessions, etc.) 

I think with the examples, it's clear.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,311 @@ Account Expires: Never</screen>
+     never be used for everyday system administration tasks.
+    </para>
+
+    <itemizedlist>
```
A procedure it will be. I think it works out well.

tomschr commented:
@mgerstner Just one question Matthias: I'd like to finish this PR by end of this week. Do you think, this is feasible?

mgerstner commented:
> @mgerstner Just one question Matthias: I'd like to finish this PR by end of this week. Do you think, this is feasible?

I will push changes in a moment. The rest depends upon how quickly we can agree on the changes.

mgerstner commented:
I've added two new commits that should address most of the changes you requested, except the few that I disputed. Please have a look and mark the things as resolved that you are happy with.

Before you merge anything please let me rebase these commits to achieve a cleaner git history.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+      One might be tempted to add the <literal>pam_securetty</literal> PAM
+      module to the <literal>/etc/pam.d/common-auth</literal> file to make it
+      effective for all PAM services. This would break <literal>su</literal>
+      and <literal>sudo</literal>, however, because these tools would then
```
Did you miss this? Or do you disagree with my change? :wink: Can't find it in your file.

mgerstner replied:
Seems I missed this. GitHub is not very helpful recently e.g. by "hiding comments" by default.

tomschr replied:
No problem, don't worry. :-) Yes, it happens sometimes to me too.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+      You can determine the device name of the terminal you are currently
+      logged into by running the <literal>tty</literal> command. Be careful
+      not to do this in a virtual terminal like via SSH or in a graphical
+      session (device names <literal>/dev/pts/&lt;N&gt;</literal>) but only
```
Same here? :wink:

mgerstner replied:
I actually adressed this one, but missed the `<replaceable>` part.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+      logged into by running the <literal>tty</literal> command. Be careful
+      not to do this in a virtual terminal like via SSH or in a graphical
+      session (device names <literal>/dev/pts/&lt;N&gt;</literal>) but only
+      from an actual login terminal reachable via Alt+F&lt;N&gt;.
```
This too. :wink:

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+      <screen>PermitRootLogin no</screen>
+     </listitem>
+     <listitem>
+      <para>Restart the <literal>sshd</literal> service to make the changes effective:</para>
```
This is also missing in the file. Or do you disagree? :wink:

mgerstner replied:
Maybe GitHub didn't send me all your comments as email, because I worked on each mail I got in my inbox. Sorry for the confusion.

tomschr replied:
All good. Don't worry. :-)

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+     <colspec colname="col_7" colwidth="14.2858*"/>
+     <thead>
+      <row>
+       <entry align="left" valign="top">package</entry>
+       <entry align="left" valign="top">shell personalities</entry>
+       <entry align="center" valign="top">shell variable</entry>
+       <entry align="center" valign="top">time unit</entry>
+       <entry align="left" valign="top">readonly setting</entry>
+       <entry align="center" valign="top">config path (only login shell)</entry>
+       <entry align="center" valign="top">config path (all shells)</entry>
```
I can't see this change as well.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,301 @@ Account Expires: Never</screen>
+       <entry align="center" valign="top"><para>seconds</para></entry>
+       <entry align="left" valign="top"><para><literal>readonly TMOUT=</literal></para></entry>
+       <entry align="left" valign="top"><para><literal>/etc/profile.local, /etc/profile.d/</literal></para></entry>
+       <entry align="left" valign="top"><para><literal>/etc/zsh.zshrc.local</literal></para></entry>
```
This too.

tomschr commented:
@mgerstner Thanks Matthias for all your efforts! Much appreciated! :+1: 

I've went through all the changes and marked them as resolved. The suggestions that are left are mostly formal issues (attributes etc). To my understanding, this shouldn't be really controversial. :wink:  Could you have a look?

Not sure if you know that or not. Sorry, I didn't want to educate you. Probably you all know that already.
Just would like to make you aware of one thing in GitHub. You can [incorporate suggestions quite easily](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request) without leaving GitHub. One benefit is they automatically mark the change as "resolved". See the link for more details. :wink:

mgerstner commented:
> Not sure if you know that or not. Sorry, I didn't want to educate you. Probably you all know that already.
> Just would like to make you aware of one thing in GitHub.

Yeah I figured that. Maybe I try that next time. I'd need to add the changes to batch commit. But I also want to keep the changes for the sections separate so I can rebase easily in the end. I can do that more easily on the command line

mgerstner commented:
So the missing stuff should be covered now. When you're fine with the current state then please tell me so I can rebase and squash the two additional commits before merging.

tomschr commented:
> Yeah I figured that. Maybe I try that next time. I'd need to add the changes to batch commit.

Sure, whatever works best for you. It wasn't my intention to force you to some workflow. :-)

> So the missing stuff should be covered now. When you're fine with the current state then please tell me so I can rebase and squash the two additional commits before merging.

Will do! Thanks! :+1:
I think, it would be a good idea to squash the commits before we hand-over to Liam for proofreading.

tomschr commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,325 @@ Account Expires: Never</screen>
   </sect1>
   <xi:include href="hardening_pam_stack.xml"/>
+  <sect1 xml:id="sec-sec-prot-restrict-root">
+   <title>Restricting &rootuser; logins</title>
```
Just one small correction:

```suggestion
   <title>Restricting &rootuser; Logins</title>
```

tomschr commented:
@mgerstner: There is a small change of "login" -> "Login" above. Should be fairly easy. :-) I think, when this is integrated, I'm fine with the current state. :+1:

Just squash all the commits, add a reference to the BSC in the commit message, and we can give it to Liam.

Thanks Matthias for your wonderful work!

mgerstner commented:
I've squashed the commits. So from my side this is ready to be merged. Expect I still overlooked something.

tomschr commented:
@mgerstner Thanks! I see 3 commits. May I ask to squash them into one? :wink:

mgerstner commented:
> @mgerstner Thanks! I see 3 commits. May I ask to squash them into one?

IMO the exercise is not to end up with as few commits as possible but to end up with well structured commits. Now we have one that harmonizes the existing hardening guide regarding itemizedlist defaults. Then one that adds the section about restricting root logins. And one that adds the section about shell inactivity timeouts.

What would we gain from squashing them into one?

tomschr commented:
> [...] What would we gain from squashing them into one?

Well, we (or better I) need to cherry-pick that in different maintenance branches too. This is easier with _one_ commit of course. :wink: 

Also, this PR is about a specific Bugzilla entry. From my point of view, I would see that in one single commit. But I know there are probably different perspectives on this. The more important thing is that I can handle it easier when cherry-picking.

mgerstner commented:
> Well, we (or better I) need to cherry-pick that in different maintenance branches too. This is easier with _one_ commit of course. 
> 
> Also, this PR is about a specific Bugzilla entry. From my point of view, I would see that in one single commit. But I know there are probably different perspectives on this. The more important thing for me is that it's easier when I do cherry-picking.

I think you are following a wrong pattern here. You shouldn't structure commits to simplify your merge workflow. This is something you do only once or twice, but the commit history will stay forever. Anybody trying to find out what was going on will have a hard time dissecting the individual parts. Also there can come up situations like "I only want that section" or "I only want the default attributes removed, but not the additional sections". In this situation the squashing will backfire.

But it is your repo, of course, so if you want it squashed, then so be it.

tomschr commented:
> I think you are following a wrong pattern here

Yeah, could be. 😁

Well, we have maintenance branches which are normally "stable". These get updates only when we cherry-pick something from master or explicitly add it into the branch. Apart from these, nothing is added.

> Also there can come up situations like "I only want that section" or "I only want the default attributes removed, but not the additional sections". In this situation the squashing will backfire.

Theoretically that's true. However, tThis happened to me only a few times over the years. Maybe for docs it's a big different than for code. :man_shrugging:  Normally, our bugs are a "all or nothing approach". :wink: 

I think, @lproven can start proofreading the work. We can later decide what we can do with our commits.

mgerstner commented:
It's fine with me, I already squashed all into one commit. Please go ahead.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,325 @@ Account Expires: Never</screen>
+   <title>Restricting &rootuser; Logins</title>
+
+   <para>
+    By default, the &rootuser; user is assigned a password
+    and can login via various methods like on any local terminal, in a
+    graphical session, or remotely via SSH. These methods should be restricted
```
We have both `local console` and `local terminal` used throughout the document currently, but not `system console` or `system terminal`. So I'd like to avoid introducing another term for the same thing.

mgerstner replied:
I will push a change without the `system console` bit.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,325 @@ Account Expires: Never</screen>
+    By default, the &rootuser; user is assigned a password
+    and can login via various methods like on any local terminal, in a
+    graphical session, or remotely via SSH. These methods should be restricted
+    as far as possible. Shared usage of the root account should be avoided,
```
It looks like this is missing a bit. When there is a fullstop the next word should start with a capital letter. I will adjust accordingly.

mgerstner commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,325 @@ Account Expires: Never</screen>
+    and can login via various methods like on any local terminal, in a
+    graphical session, or remotely via SSH. These methods should be restricted
+    as far as possible. Shared usage of the root account should be avoided,
+    instead individual administrators should use tools like
```
Ah there is the bit that I was missing before. Sorry.

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    <note>
+     <para>
+      The steps shown here are tailored towards PC architectures (&x86; and
+      &x86-64;). On architectures like &power; different terminal
```
```suggestion
      &x86-64;). On architectures such as &power;, different terminal
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+      lock yourself out completely by specifying wrong terminal device names.
+      You can determine the device name of the terminal you are currently
+      logged into by running the <literal>tty</literal> command. Be careful
+      not to do this in a virtual terminal like via SSH or in a graphical
```
```suggestion
      not to do this in a virtual terminal, such as via SSH or in a graphical
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+      You can determine the device name of the terminal you are currently
+      logged into by running the <literal>tty</literal> command. Be careful
+      not to do this in a virtual terminal like via SSH or in a graphical
+      session (device names <filename>/dev/pts/<replaceable>N</replaceable></filename>)
```
```suggestion
      session (device names <filename>/dev/pts/<replaceable>N</replaceable></filename>),
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+ auth     [user_unknown=ignore success=ok ignore=ignore auth_err=die default=bad] pam_securetty.so noconsole
+ auth     include        common-auth</screen>
+      <para>
+        This will include the <literal>pam_securetty</literal> PAM module during the
```
```suggestion
        This will include the <literal>pam_securetty</literal> module during the
```

lproven replied:
Since the "M" in "PAM" already stands for "module", this is redundant.

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+ auth     include        common-auth</screen>
+      <para>
+        This will include the <literal>pam_securetty</literal> PAM module during the
+        authentication process on local consoles to allow root only to login on
+        tty devices that are listed in the file <filename>/etc/securetty</filename>.
```
```suggestion
        authentication process on local consoles, which restricts &rootuser; to logging-in
        only on TTY devices that are listed in the file <filename>/etc/securetty</filename>.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+      <para>
+       Check whether logins to other terminals will be rejected for
+       &rootuser;. A login on <literal>tty2</literal>, for example, should be
+       rejected right away without even the root password being queried.
```
```suggestion
       rejected immediately, without even querying the account password.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+
+    <important>
+     <para>
+      Avoid adding the <literal>pam_securetty</literal> PAM module to the
+      <filename>/etc/pam.d/common-auth</filename> file. This change would
+      break <command>su</command> and <command>sudo</command>, however,
+      because these tools would then also reject &rootuser; authentications.
```
```suggestion
      Do not add the <literal>pam_securetty</literal> module to the
      <filename>/etc/pam.d/common-auth</filename> file. This would
      break the <command>su</command> and <command>sudo</command> commands,
      because these tools would then also reject &rootuser; authentications.
```

lproven replied:
This might be better as a [warning] -- do you agree, @tomschr?

tomschr replied:
Yes, we can do that in the fine-tuning part. :wink:

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    <important>
+     <para>
+      These configuration changes will also cause root logins on serial
+      consoles like <filename>/dev/ttyS0</filename> to be denied. In case you
```
```suggestion
      consoles such as <filename>/dev/ttyS0</filename> to be denied. In case you
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    <important>
+     <para>
+      These configuration changes will also cause root logins on serial
+      consoles like <filename>/dev/ttyS0</filename> to be denied. In case you
+      require such use cases you need to list the respective tty devices
+      additionally in the <filename>/etc/securetty</filename> file.
```
```suggestion
      consoles like <filename>/dev/ttyS0</filename> to be denied. If you
      require such use cases, then also list the respective TTY devices
      in the <filename>/etc/securetty</filename> file.
```

lproven replied:
Github timeouts again. >_<  I cannot revise my previous suggestion.

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    <title>Restricting Graphical Session Logins</title>
+
+    <para>
+     To improve security on your server, avoid to use graphical user
```
```suggestion
     To improve security on your server, avoid using graphical
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+     not designed to be run as &rootuser; and are more likely to
+     contain security issues than console programs. If you require a graphical
+     login, use a non-&rootuser; login. Configure your system to disallow
+     &rootuser; login on graphical session.
```
```suggestion
     &rootuser; from logging in to graphical sessions.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    </para>
+
+    <para>
+     To deny the &rootuser; user login to graphical sessions you
+     can apply the same basic steps as outlined in
+     <xref linkend="sec-sec-prot-restrict-root-tty"/>.
+     You only need to also add the <literal>pam_securetty</literal> module to
+     the PAM stack file belonging to the display manager, for example
+     <filename>/etc/pam.d/gdm</filename> for GDM. The graphical session also
+     runs on a TTY device, <literal>tty7</literal>, by default. Therefore,
+     when you restrict &rootuser; logins to
+     <literal>tty1</literal>, then a &rootuser; login in the
+     graphical session will be denied.
```
```suggestion
     To prevent &rootuser; from logging in to graphical sessions, you
     can apply the same basic steps as outlined in
     <xref linkend="sec-sec-prot-restrict-root-tty"/>.
     Just add the <literal>pam_securetty</literal> module to the PAM
     stack file belonging to the display manager&mdash;for example,
     <filename>/etc/pam.d/gdm</filename> for GDM. The graphical
     session also runs on a TTY device: by default, <literal>tty7</literal>.
     Therefore, if you restrict &rootuser; logins to <literal>tty1</literal>,
     then &rootuser; will be denied login in the graphical session.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   <sect2 xml:id="sec-sec-prot-restrict-root-ssh">
+    <title>Restricting SSH Logins</title>
+    <para>
+     By default the &rootuser; user is also allowed to login
+     into a machine remotely via the SSH network protocol (if the SSH port is
+     not blocked by the firewall). To restrict this possibility the following
+     change of the OpenSSH configuration needs to be performed:
```
```suggestion
     By default, the &rootuser; user is also allowed to log
     into a machine remotely via the SSH network protocol (if the SSH port is
     not blocked by the firewall). To restrict this, make the following
     change to the OpenSSH configuration:
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    </procedure>
+    <note>
+     <para>
+      Using the PAM module <literal>pam_securetty</literal> is not suitable in
+      case of OpenSSH, because SSH logins not necessarily go through the PAM
+      stack during the auth phase (e.g. when using SSH public key authentication).
+      Also an attacker could be able to differentiate between a wrong password
+      and a successful login that was only rejected later on by policy.
```
```suggestion
      Using the PAM <literal>pam_securetty</literal> module is not suitable in
      case of OpenSSH, because not all SSH logins go through the PAM
      stack during authorization (for example, when using SSH public-key authentication).
      In addition, an attacker could differentiate between a wrong password
      and a successful login that was only rejected later on by policy.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   <title>Setting an Inactivity Timeout for Interactive Shell Sessions</title>
+   <para>
+    It can be a good idea to terminate an interactive shell session after a
+    certain time of inactivity, for example to:
```
```suggestion
    certain period of inactivity; for example, to:
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+       avoid waste of system resources.
+      </para>
+     </listitem>
+    </itemizedlist>
```
I think an itemizedlist is not needed for just 2 alternatives, but it is not a serious problem.

tomschr replied:
I've used inline text instead of the 2 items, comitted in 18deb0107.

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    </itemizedlist>
+   <para>
+    By default, there is no inactivity timeout for shells. Nothing will
+    happen if a shell stays open unused for days or even years. It is possible
```
```suggestion
    happen if a shell stays open and unused for days or even years. It is possible
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+    common types of shells.
+   </para>
+   <para>
+    The inactivity timeout can be configured for login shells only or for all
+    interactive shells. In the latter case the inactivity timeout is running
+    individually for each shell instance. This means that timeouts will
+    accumulate. When a sub or child shell is started then the timeout runs for
+    the sub or child shell and only afterwards will the
+    timeout of the parent continue running.
```
```suggestion
    The inactivity timeout can be configured for login shells only or for all
    interactive shells. In the latter case the inactivity timeout is running
    individually for each shell instance. This means that timeouts will
    accumulate. When a sub- or child-shell is started, then a new timeout begins for
    the sub- or child-shell, and only afterwards will the
    timeout of the parent continue running.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   </para>
+   <para>
+    The following table contains configuration details for a selection of
+    common shells shipped with SLE:
```
I suspect that there is an entity for "SLE" here.

tomschr replied:
I used the `&productname;` in commit 18deb0107

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   </informaltable>
+   <para>
+    Every listed shell supports an internal timeout shell variable that can be
+    set to a specific time value to activate the inactivity timeout. If you want
```
```suggestion
    set to a specific time value to cause the inactivity timeout. If you want
```

lproven replied:
"Activate the inactivity" sounded odd to me.

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   <para>
+    Every listed shell supports an internal timeout shell variable that can be
+    set to a specific time value to activate the inactivity timeout. If you want
+    to prevent users from overriding the timeout setting you can mark the
```
```suggestion
    to prevent users from overriding the timeout setting, you can mark the
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+
+   <note>
+    <para>
+     This feature is only a utility that can help avoiding risks if a user is
+     forgetful or follows unsafe practices. It does not protect against
+     malicious users, however. The timeout discussed here only applies to
+     certain interactive wait states of a shell. A malicious user can always
+     find ways to circumvent the timeout and keep their session open anyway.
```
```suggestion
     This feature is only helpful for avoiding risks if a user is forgetful or follows
     unsafe practices. It does not protect against hostile users. The timeout
     only applies to interactive wait states of a shell. A malicious user can always
     find ways to circumvent the timeout and keep their session open regardless.
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+   </note>
+
+   <para>
+    To configure the inactivity timeout you need to add the matching timeout
+    variable declaration to a shell startup script. Use either the path for
+    login shells only, or the one for all shells as listed in the table. The
+    following example uses paths and settings that are suitable for
+    <command>bash</command> and <command>ksh</command>
+    to setup a readonly login shell timeout that cannot be overriden by users.
+    Create the file <filename>/etc/profile.d/timeout.sh</filename> with the
+    following content:
```
```suggestion
    To configure the inactivity timeout, you need to add the matching timeout
    variable declaration to each shell's startup script. Use either the path for
    login shells only, or the one for all shells, as listed in the table. The
    following example uses paths and settings that are suitable for
    <command>bash</command> and <command>ksh</command>
    to setup a read-only login shell timeout that cannot be overridden by users.
    Create the file <filename>/etc/profile.d/timeout.sh</filename> with the
    following content:
```

lproven commented on xml/hardening.xml:
```diff
@@ -3445,6 +3445,327 @@ Account Expires: Never</screen>
+
+   <tip>
+    <para>
+     We recommend to use the <command>screen</command> tool in order to
+     detach sessions before logging out. Screen sessions are not terminated
+     and can be re-attached whenever it is required. Active session can be
+     locked without logging out (read about
```
```suggestion
     We recommend using the <command>screen</command> tool in order to
     detach sessions before logging out. <command>screen</command> sessions
     are not terminated and can be re-attached whenever required. An active
     session can be locked without logging out (read about
```

tomschr commented:
For some reasons I don't know, GitHub doesn't allow me to do batch committing of Liam's changes. So I need to apply them individually. :-(
